test/metabase/test/test_data.clj

```diff
@@ -0,0 +1,205 @@
+     [:map {:closed true}
+      [:native ::lib.schema.common/non-blank-string]]
+     ::lib.schema.common/base-type])
+
+(mr/def ::schema.field
+  [:map {:closed true}
+   [:name      ::lib.schema.common/non-blank-string]
+   [:base-type ::schema.field.base-type]
+   ;;
+   ;; field modifiers
+   ;;
+   [:not-null?         {:optional true} [:maybe :boolean]] ; default is nullable
+   [:unique?           {:optional true} [:maybe :boolean]] ; implied for PK fields, so doesn't need to be set there.
+   [:pk?               {:optional true} [:maybe :boolean]]
+   [:fk                {:optional true} [:maybe [:tuple
+                                                 #_table-name ::lib.schema.common/non-blank-string
+                                                 #_field-name ::lib.schema.common/non-blank-string]]]
+   [:comment           {:optional true} [:maybe ::lib.schema.common/non-blank-string]]
+   ;; Should we manually create a separate/additional index for this field? This should be falsey for PK and FK fields
+   ;; since they normally get an index automatically.
+   [:indexed?          {:optional true} [:maybe :boolean]]
+   ;;
+   ;; these are only important for sync, driver implementations can ignore these columns.
+   ;;
+   [:semantic-type     {:optional true} [:maybe ::lib.schema.common/semantic-or-relation-type]]
+   [:effective-type    {:optional true} [:maybe ::lib.schema.common/base-type]]
+   [:coercion-strategy {:optional true} [:maybe ms/CoercionStrategy]]
+   [:visibility-type   {:optional true} [:maybe (into [:enum] field/visibility-types)]]])
+
+(mr/def ::schema.table
+  [:map {:closed true}
+   [:name   ::lib.schema.common/non-blank-string]
+   [:fields [:sequential ::schema.field]]
+   [:comment {:optional true} [:maybe ::lib.schema.common/non-blank-string]]])
+
+(mr/def ::schema.database
+  [:map
+   [:tables [:sequential ::schema.table]]])
+
+(mr/def ::reducible-rows
+  (ms/InstanceOfClass clojure.lang.IReduceInit))
+
+(mr/def ::artifact
+  [:map
+   [:filename   {:optional true} [:maybe ::lib.schema.common/non-blank-string]]
+   [:statements {:optional true} [:maybe [:sequential some?]]]])
+
+(mr/def ::artifacts
+  [:map {:closed true}
+   [:server {:optional true} [:maybe [:sequential ::artifact]]]
+   [:db     {:optional true} [:maybe [:sequential ::artifact]]]])
+
+(mu/defn ^:private reducible-edn-file :- (ms/InstanceOfClass clojure.lang.IReduceInit)
+  [filename :- :string]
+  (when-not (.exists (io/file filename))
+    (throw (ex-info "File does not exist" {:filename filename})))
+  (reify clojure.lang.IReduceInit
+    (reduce [_this rf init]
+      (with-open [r (clojure.tools.reader.reader-types/push-back-reader (java.io.FileReader. (io/file filename)))]
+        (loop [acc init]
+          (if (reduced? acc)
+            (unreduced acc)
+            (if-let [object (edn/read {:eof nil, :readers {'t u.date/parse}} r)]
+              (recur (rf acc object))
+              acc)))))))
```

Comment:
Instead of reading massive files like `test-data.edn` (53,000 lines, 2.2 MB) into memory and parsing them into various record types and caching them in memory forever (a gigantic waste of memory) we now have highly optimized code that reads data out from EDN only when needed in a streaming/reducible manner and generates relevant CSV files in a streaming/reducible manner. 

I haven't measured the potential impact on test memory usage but it has to definitely shave off a lot of usage.